conda/notices/core.py

```diff
@@ -35,69 +36,104 @@ def display_notices(
         viewed_notices = cache.get_viewed_channel_notice_ids(cache_file, channel_notices)
-        num_viewed_notices = len(viewed_notices)
+        viewed_channel_notices = len(viewed_notices)
 
     channel_notices = filter_notices(channel_notices, limit=limit, exclude=viewed_notices)
-    if len(channel_notices) == 0:
-        return
 
-    views.print_notices(channel_notices)
+    return ChannelNoticeResultSet(
+        channel_notices=channel_notices,
+        viewed_channel_notices=viewed_channel_notices,
+        total_number_channel_notices=total_number_channel_notices,
+    )
+
+
+def display_notices(channel_notice_set: ChannelNoticeResultSet) -> None:
+    """
+    Prints the channel notices to std out
+    """
+    views.print_notices(channel_notice_set.channel_notices)
 
     # Updates cache database, marking displayed notices as "viewed"
-    cache.mark_channel_notices_as_viewed(cache_file, channel_notices)
+    cache_file = cache.get_notices_cache_file()
+    cache.mark_channel_notices_as_viewed(cache_file, channel_notice_set.channel_notices)
 
-    views.print_more_notices_message(num_total_notices, len(channel_notices), num_viewed_notices)
+    views.print_more_notices_message(
+        channel_notice_set.total_number_channel_notices,
+        len(channel_notice_set.channel_notices),
+        channel_notice_set.viewed_channel_notices,
+    )
 
 
 def notices(func):
     """
     Wrapper for "execute" entry points for subcommands.
 
+    If channel notices need to be fetched, we do that first and then
+    run the command normally. We then display these notices at the very
+    end of the command output so that the user is more likely to see them.
+
+    This ordering was specifically done to address the following bug report:
+        - https://github.com/conda/conda/issues/11847
+
     Args:
         func: Function to be decorated
     """
-
     @wraps(func)
     def wrapper(*args, **kwargs):
-        return_value = func(*args, **kwargs)
-
-        if is_channel_notices_enabled(context):
-            display_notices(
+        if is_channel_notices_enabled(context) and is_channel_notices_cache_expired():
+            channel_notice_set = retrieve_notices(
                 limit=context.number_channel_notices,
                 always_show_viewed=False,
                 silent=True,
             )
+            return_value = func(*args, **kwargs)
+            display_notices(channel_notice_set)
+
+            return return_value
 
-        return return_value
+        else:
+            return func(*args, **kwargs)
 
     return wrapper
 
 
 def get_channel_name_and_urls(
-    channels: [Sequence[Channel]],
-) -> Sequence[Tuple[ChannelUrl, ChannelName]]:
-    """Return a sequence of Channel URL and name"""
-
-    def ensure_endswith(value: str, ends: str) -> str:
-        return value if value.endswith(ends) else f"{value}{ends}"
+    channels: Sequence[Union[Channel, MultiChannel]],
+) -> Sequence[tuple[ChannelUrl, ChannelName]]:
```

Comment:
Good practice for the function input types to be generic but for the output type to be explicit:

```suggestion
) -> list[tuple[ChannelUrl, ChannelName]]:
```